pkg/kubelet/dockertools/manager.go

```diff
@@ -1119,70 +1135,81 @@ func (dm *DockerManager) KillPod(pod kubecontainer.Pod) error {
-	return dm.killContainer(containerID)
+	case container == nil || pod == nil:
+		// Read information about the container from labels
+		inspect, err := dm.client.InspectContainer(string(containerID))
+		if err != nil {
+			return err
+		}
+		storedPod, storedContainer, cerr := containerAndPodFromLabels(inspect)
+		if cerr != nil {
+			glog.Errorf("unable to access pod data from container: %v", err)
+		}
+		if container == nil {
+			container = storedContainer
+		}
+		if pod == nil {
+			pod = storedPod
+		}
+	}
+	return dm.killContainer(containerID, container, pod)
 }
 
-func (dm *DockerManager) killContainer(containerID types.UID) error {
+// killContainer accepts a containerID and an optional container or pod containing shutdown policies. Invoke
+// KillContainerInPod if information must be retrieved first.
+func (dm *DockerManager) killContainer(containerID types.UID, container *api.Container, pod *api.Pod) error {
 	ID := string(containerID)
-	glog.V(2).Infof("Killing container with id %q", ID)
-	inspect, err := dm.client.InspectContainer(ID)
-	if err != nil {
-		return err
+	name := ID
+	if container != nil {
+		name = fmt.Sprintf("%s %s", name, container.Name)
 	}
-	var found bool
-	var preStop string
-	if inspect != nil && inspect.Config != nil && inspect.Config.Labels != nil {
-		preStop, found = inspect.Config.Labels[kubernetesPodLabel]
+	if pod != nil {
+		name = fmt.Sprintf("%s %s/%s", name, pod.Namespace, pod.Name)
 	}
-	if found {
-		var pod api.Pod
-		err := latest.Codec.DecodeInto([]byte(preStop), &pod)
-		if err != nil {
-			glog.Errorf("Failed to decode prestop: %s, %s", preStop, ID)
-		} else {
-			name := inspect.Config.Labels[kubernetesContainerLabel]
-			var container *api.Container
-			for ix := range pod.Spec.Containers {
-				if pod.Spec.Containers[ix].Name == name {
-					container = &pod.Spec.Containers[ix]
-					break
-				}
-			}
-			if container != nil {
-				glog.V(1).Infof("Running preStop hook")
-				if err := dm.runner.Run(ID, &pod, container, container.Lifecycle.PreStop); err != nil {
-					glog.Errorf("failed to run preStop hook: %v", err)
-				}
-			} else {
-				glog.Errorf("unable to find container %v, %s", pod, name)
-			}
+
+	gracePeriod := uint(minimumGracePeriodInSeconds)
+	if pod != nil && pod.DeletionGracePeriodSeconds != nil {
+		gracePeriod = uint(*pod.DeletionGracePeriodSeconds)
+	}
+	glog.V(2).Infof("Killing container %q with %d grace period", name, gracePeriod)
+
+	if pod != nil && container != nil && container.Lifecycle != nil && container.Lifecycle.PreStop != nil {
+		glog.V(4).Infof("Running preStop hook for container %q", name)
+		start := util.Now()
+		// TODO: timebox PreStop execution to at most gracePeriod
+		if err := dm.runner.Run(ID, pod, container, container.Lifecycle.PreStop); err != nil {
+			glog.Errorf("preStop hook for container %q failed: %v", name, err)
 		}
+		gracePeriod -= uint(util.Now().Sub(start.Time).Seconds())
```

Comment:
It was because of the signature of StopContainer, I'll use int64 and and simply drop anything out of uint()

----- Original Message -----
> >  		}
> > +		gracePeriod -= uint(util.Now().Sub(start.Time).Seconds())
> 
> Does this saturate at 0? Signed int (int64) would be better.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/8641/files#r31465110